File: bank(csa2)/bankClient/src/BankClient.java
import com.jbosak.csa.bank.server.Session;
import com.jbosak.csa.bank.server.User;
import org.apache.xmlrpc.XmlRpcException;
import org.apache.xmlrpc.client.XmlRpcClient;
import org.apache.xmlrpc.client.XmlRpcClientConfig;
import org.apache.xmlrpc.client.XmlRpcClientConfigImpl;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.Scanner;
import java.util.Vector;

class BankClient {
    private static final String SERVER_NAME = "Bank";
    private static final String HOST_URL = "http://localhost:5059";

    private XmlRpcClient client;
    private Session userSession;

    BankClient() throws MalformedURLException {
        this.client = new XmlRpcClient();


        XmlRpcClientConfigImpl config = new XmlRpcClientConfigImpl();
        config.setEnabledForExceptions(true);
        config.setEnabledForExtensions(true);

        client.setTypeFactory(new SessionFactory(client));
        client.setConfig(config);
        config.setServerURL(new URL(HOST_URL));
    }

    void options(BankCallback callback) throws XmlRpcException {
        Vector<Object> params  = new Vector<Object>();
//        if (userSession != null){
//            params.add(userSession.getUser().getUsername());
//            params.add(userSession.getUser().getPassword());
//
//        }
                client.executeAsync(SERVER_NAME + ".options", params, callback);
    }

    void login(Scanner scanner, BankCallback callback) throws XmlRpcException {
        Vector<String> user  = new Vector<String>();
        user.add(scanner.next());
        user.add(scanner.next());

        client.executeAsync(SERVER_NAME + ".login", user, callback);
    }

    public void register(Scanner scanner, BankCallback callback) throws XmlRpcException {
        System.out.println("register");
        Vector<Object> user  = new Vector<Object>();
        user.add(scanner.next());
        user.add(scanner.next());
        user.add(Integer.valueOf(scanner.next()));
        client.executeAsync(SERVER_NAME + ".register", user, callback);
    }

    public void handleLoggedIn(User user) {
        if (user != null){
            XmlRpcClientConfigImpl config = (XmlRpcClientConfigImpl) client.getConfig();
            config.setBasicPassword(user.getPassword());
            config.setBasicUserName(user.getUsername());
            client.setConfig(config);
            System.out.println("Logged in successfully");
        }else{
            userSession = null;
            System.out.println("Wrong login or password");
        }

    }

    public void accountBalance(BankCallback callback) throws XmlRpcException {
        Vector<Object> params  = new Vector<Object>();
        params.add(((XmlRpcClientConfigImpl)client.getConfig()).getBasicUserName());
        client.executeAsync(SERVER_NAME + ".accountBalance", params, callback);
        System.out.println("AccountBalance");
    }

    public void transfer(Scanner scanner, BankCallback callback) throws XmlRpcException {
        Vector<Object> transfer  = new Vector<Object>();
        transfer.add(scanner.next());
        transfer.add(Integer.valueOf(scanner.next()));
        client.executeAsync(SERVER_NAME + ".transfer", transfer, callback);
    }

    public void logout(BankCallback callback) throws XmlRpcException {
        Vector<Object> params  = new Vector<Object>();
        client.executeAsync(SERVER_NAME + ".logout", params, callback);

    }

    public void anotherMethod(Scanner scanner) {
        Vector<User> emptyVector = new Vector<User>();
        User user = new User("xd","aa");
        System.out.println(user);
//        emptyVector.add(user);
        User response = null;
        try {
//            response = (User)
            System.out.println(client.execute(SERVER_NAME + "." + scanner.next(), emptyVector));
        } catch (XmlRpcException e) {
            System.out.println(e);
            e.printStackTrace();
        }

//        assert response != null;
//        System.out.println(response);
    }

    public void handleLogout(Boolean success) {
        if (success){
            XmlRpcClientConfigImpl config = (XmlRpcClientConfigImpl) client.getConfig();
            config.setBasicPassword(null);
            config.setBasicUserName(null);
            client.setConfig(config);
            System.out.println("Logged out successfully");

        }
    }
}

File: bank(csa2)/bankClient/src/BankClient.kt
import com.jbosak.csa.bank.server.UserManager
import com.jbosak.csa.bank.server.model.User
import org.apache.xmlrpc.client.XmlRpcClientConfigImpl
import java.lang.reflect.UndeclaredThrowableException
import java.net.MalformedURLException
import java.net.URL
import java.util.*


class BankClient {

    private var clientProxy: ClientProxy? = null

    private val userManager: UserManager

    private var config: XmlRpcClientConfigImpl

    init {

        try {
            clientProxy = ClientProxy(URL("$HOST_URL"))
        } catch (e: MalformedURLException) {
            e.printStackTrace()
        }
        var client = clientProxy!!.xmlRpcClient

        config = XmlRpcClientConfigImpl()
        config.serverURL = URL("$HOST_URL/$SERVER_NAME")
        config.isEnabledForExtensions = true
        config.isContentLengthOptional = true
        config.isEnabledForExceptions = true

        clientProxy?.config = config
        userManager = clientProxy?.wrap(UserManager::class.java, SERVER_NAME)!!

        client.setConfig(config)

    }

    fun options() {
        UserManager::class.java.methods.forEach { print( it.name + ", ") }
        println()
    }

    fun login(scanner: Scanner) {
        val user = userManager.login(scanner.next(), scanner.next())
        if (user == null) println("Something went wrong")
        else {
            println("Logged in successfully")
            config.basicUserName = user.username
            config.basicPassword = user.password
        }
    }

    fun register(scanner: Scanner) {
        val user = User(scanner.next(),scanner.next(), Integer.valueOf(scanner.next()))
        val success = userManager.register(user)
        if (success) println("Account created")
        else println("User with username: ${user.username} already exists")
    }


    fun accountBalance() {
        var username  = config.basicUserName
        if (username == null) username = ""
        try {
            println("Your amount of money: " + userManager.accountBalance(username))

        }catch (e: UndeclaredThrowableException){
            println("You're not logged in!")
        }
    }

    fun transfer(scanner: Scanner) {

        val config:XmlRpcClientConfigImpl? = clientProxy?.xmlRpcClient?.config as XmlRpcClientConfigImpl
        if (config == null || config.basicUserName.isNullOrEmpty()){
            println("You're not logged in!")
            return
        }else{
            val success = userManager.transfer(config.basicUserName, scanner.next(), scanner.nextInt())
            if (success) println("transfer successfully finished")
            else println("Something went wrong")

        }

    }

    fun logout() {
        val config = clientProxy?.xmlRpcClient?.clientConfig as XmlRpcClientConfigImpl
        if (config.basicUserName.isNullOrEmpty()){
            println("You're not even logged in!")
            return
        }
        userManager.logout()
        println("logged out")
        config.basicUserName = null
        config.basicPassword = null

    }

    companion object {
        private const val SERVER_NAME = "UserManager"
        private const val HOST_URL = "http://localhost:5059"
    }
}
